cli/daemon/httpserver.go

```diff
@@ -0,0 +1,145 @@
+package daemon
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/tarantool/tt/cli/daemon/api"
+	"github.com/tarantool/tt/cli/ttlog"
+)
+
+const (
+	defaultStopTimeout = 100 * time.Millisecond
+)
+
+type HTTPServer struct {
+	// listenIface is a network interface the IP address
+	// should be found on to bind http server socket.
+	listenIface string
+	// port is a port number to be used for http server.
+	port int
+	// srv is http.Server instance.
+	srv *http.Server
+	// timeout is the time that was provided
+	// to the HTTP server to shutdown correctly.
+	timeout time.Duration
+	// logger is  a log file the HTTP server will write to.
+	logger *ttlog.Logger
+}
+
+// listenIP discovers IP address on the specified interface.
+// If interface name is not specified returned value is 0.0.0.0;
+// returns the first discovered IP address on the specified interface.
+func (httpServer *HTTPServer) listenIP() (string, error) {
+	if httpServer.listenIface == "" {
+		return net.IPv4zero.String(), nil
+	}
+
+	iface, err := net.InterfaceByName(httpServer.listenIface)
+	if err != nil {
+		return "", err
+	}
+
+	if iface.Flags&net.FlagUp == 0 {
+		return "", fmt.Errorf("Interface down")
+	}
+
+	addrs, err := iface.Addrs()
+	if err != nil {
+		return "", err
+	}
+
+	// The socket will bind to the first discovered
+	// IP address on the specified interface.
+	for _, addr := range addrs {
+		var ip net.IP
+		switch v := addr.(type) {
+		case *net.IPNet:
+			ip = v.IP
+		case *net.IPAddr:
+			ip = v.IP
+		}
+
+		if ip == nil {
+			continue
+		}
+
+		// Not an ipv4 address.
+		if ip.To4() == nil {
```

Comment:
Why don't we work with IPv6?

Reply:
Some systems will not work correct with IPv6, but I agree that we should support IPv6 as well as IPv4. My suggestion is use IPv4 by default and use IPv6 by option. Should we fix it in this version of `tt daemon` or in separate PR in the future?